Mathlib/RingTheory/MvPowerSeries/Substitution.lean

```diff
@@ -0,0 +1,435 @@
+/-
+Copyright (c) 2024 Antoine Chambert-Loir, María Inés de Frutos Fernández. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Antoine Chambert-Loir, María Inés de Frutos Fernández
+-/
+
+import Mathlib.RingTheory.MvPowerSeries.Evaluation
+import Mathlib.RingTheory.MvPowerSeries.LinearTopology
+import Mathlib.RingTheory.PowerSeries.Basic
+import Mathlib.RingTheory.Nilpotent.Basic
+import Mathlib.Topology.UniformSpace.DiscreteUniformity
+
+/-! # Substitutions in multivariate power series
+
+Here we define the substitution of power series into other power series.
+We follow [Bourbaki, Algebra II, chap. 4, §4, n° 3][bourbaki1981]
+who present substitution of power series as an application of evaluation.
+
+As an application, we define `MvPowerSeries.rescale` which rescales a multivariate
+power series `f : MvPowerSeries σ R` by a map `a : σ → R`.
+
+## Implementation note
+
+Evaluation of a power series at adequate elements has been defined
+in `Mathlib.RingTheory.MvPowerSeries.Evaluation`.
+The goal here is to check the relevant hypotheses:
+* The ring of coefficients is endowed the discrete topology.
+* The main condition rewrites as having vanishing constant coefficient
+* Multivariate power series have a linear topology
+
+To this aim, we add a local instance
+  `localinstanceUniformSpace_bot : UniformSpace S = ⊥`
+and two associated `ContinuousSMul` local instances.
+
+## TODO
+
+* Refactor `PowerSeries.rescale` using this API.
+
+-/
+
+namespace MvPowerSeries
+
+variable {σ : Type*}
+  {A : Type*} [CommSemiring A]
+  {R : Type*} [CommRing R] [Algebra A R]
+  {τ : Type*}
+  {S : Type*} [CommRing S] [Algebra A S] [Algebra R S] [IsScalarTower A R S]
+
+open WithPiTopology
+
+/-- In this file, rings are given the discrete uniformity -/
+local instance localInstanceUniformSpace_bot : UniformSpace S := ⊥
+
+local instance : ContinuousSMul R S := DiscreteTopology.instContinuousSMul R S
+
+local instance : ContinuousSMul R (MvPowerSeries τ S) := IsScalarTower.continuousSMul S
+
+/-- Families of power series which can be substituted -/
+structure HasSubst (a : σ → MvPowerSeries τ S) : Prop where
+  const_coeff s : IsNilpotent (constantCoeff τ S (a s))
+  coeff_zero d : {s | (a s).coeff S d ≠ 0}.Finite
+
+variable {a : σ → MvPowerSeries τ S}
+
+lemma coeff_zero_iff :
+    Filter.Tendsto a Filter.cofinite (@nhds _ (@instTopologicalSpace τ S ⊥) 0) ↔
+      ∀ d : τ →₀ ℕ, {s | (a s).coeff S d ≠ 0}.Finite := by
+  simp_rw [tendsto_iff_coeff_tendsto, coeff_zero]
+  apply forall_congr'
+  intro d
+  simp [nhds_discrete]
+
+/-- A multivariate power series can be substituted if and only if
+it can be evaluated when the topology on the coefficients of its coefficients
+is the discrete topology. -/
+/- We make the arguments of `HasEval` explicit to be sure that Lean infers the correct topology. -/
```

Comment:
```suggestion
it can be evaluated when the topology on the coefficients ring
is the discrete topology. -/
/- We make the arguments of `HasEval` explicit to be sure that Lean infers the correct topology. -/
```